src/Codec/CBOR/Cuddle/CBOR/Validator.hs

```diff
@@ -0,0 +1,986 @@
+        (Literal (Value (VNInt (fromIntegral -> n)) _), Literal (Value (VUInt (fromIntegral -> m)) _)) -> -n <= T.length bs && range bound (T.length bs) m
+        (Literal (Value (VNInt (fromIntegral -> n)) _), Literal (Value (VNInt (fromIntegral -> m)) _)) -> -n <= T.length bs && range bound (T.length bs) (-m)
+controlText s Regexp ctrl =
+  getRule ctrl <&> \case
+    Literal (Value (VText rxp) _) -> case s =~ rxp :: (T.Text, T.Text, T.Text) of
+      ("", s', "") -> s == s'
+
+--------------------------------------------------------------------------------
+-- Tagged values
+
+-- | Validating a `TTagged`
+validateTagged ::
+  (MonadError (ValidationTree CDDLFail) m, MonadReader CDDL m) =>
+  Word64 -> Term -> Rule -> m (Term -> ValidationTree Term)
+validateTagged tag term rule = do
+  getRule rule >>= \case
+    Postlude PTAny -> pure (ValLeaf rule)
+    Tag tag' rule' ->
+      -- If the tag does not match, this is a direct fail
+      if tag == tag'
+        then
+          ask >>= \cddl ->
+            case runReader (runExceptT (validateTerm term rule')) cddl of
+              Left vt -> throwError (ValContainer rule [] [fmap snd vt])
+              Right vt -> pure (const $ ValContainer rule [] [vt])
+        else throwError (ValLeaf rule DirectFail)
+    Choice opts -> validateChoice (validateTagged tag term) rule opts
+    _ -> throwError (ValLeaf rule UnapplicableRule)
+
+--------------------------------------------------------------------------------
+-- Collection helpers
+
+-- | Groups might contain enums, or unwraps inside. This resolves all those to
+-- the top level of the group.
+flattenGroup :: CDDL -> [Rule] -> [Rule]
+flattenGroup cddl nodes =
+  mconcat
+    [ case resolveIfRef cddl rule of
+        Literal{} -> [rule]
+        Postlude {} -> [rule]
+        Map {} -> [rule]
+        Array {} -> [rule]
+        Choice {} -> [rule]
+        KV {} -> [rule]
+        Occur {} -> [rule]
+        Range {} -> [rule]
+        Control {} -> [rule]
+        Enum e -> case resolveIfRef cddl e of
+          Group g -> flattenGroup cddl g
+          _ -> error "Malformed cddl"
+        Unwrap g -> case resolveIfRef cddl g of
+          Map n -> flattenGroup cddl n
+          Array n -> flattenGroup cddl n
+          Tag _ n -> [n]
+          _ -> error "Malformed cddl"
+        Tag {} -> [rule]
+        Group g -> flattenGroup cddl g
+    | rule <- nodes
+    ]
+
+-- | Expand rules to reach exactly the wanted length, which must be the number
+-- of items in the container. For example, if we want to validate 3 elements,
+-- and we have the following CDDL:
+--
+-- > a = [* int, * bool]
+--
+-- this will be expanded to `[int, int, int], [int, int, bool], [int, bool,
+-- bool], [bool, bool, bool]`.
+--
+-- Essentially the rules we will parse is the choice among the expansions of the
+-- original rules.
+expandRules :: Int -> [Rule] -> Reader CDDL [[Rule]]
```

Comment:
This seems quite inefficient, it might struggle with PParams.

Reply:
Oh, never mind actually, reading it again I realized it doesn't backtrack as much as I first thought.